### ISSUE TITLE
Shared initialized firebase App instance

### ISSUE THREAD

**Issue by tfesslerCWC** [benwinding/react-admin-firebase#25]
I am integrating react-filepond-firebase into react-admin-firebase and to share the initialized firebase.app from the react-admin-firebase library.

One approach would be to pass an initialized firebase object from the client layer that's setup in its own file, such as the example below:

store.js
> import firebase from 'firebase'
> 
> const config = {
>   apiKey: "sadasdasasdasdasdasdasd-cqUU37c4",
>   authDomain: "asdasdasdasdasd.firebaseapp.com",
>   databaseURL: "https://asdasdasdasdas.firebaseio.com",
>   projectId: "asdasdasdasda",
>   storageBucket: "dadasdsadasdas.appspot.com",
>   messagingSenderId: "3423423423432432",
>   appId: "1:2343242342342:web:4234234234234"
>   };
> 
>   const MyStore = firebase.initializeApp(config);
> 
>   export default MyStore;

This file could be imported and MyStore could be passed to the React-Admin-Firebase Library, allowing additional access to the firebase app instance for realtime database and storage access.

- or - 

The initialized firebase app object that's in the react-admin-firebase library could be included and accessed  after it's initialized. 

Any assistance would be appreciated.

**Comment by benwinding** [comment]
That's a good idea @tfesslerCWC,

We could put an option to include the initialised firebase app instance into the data provider options. As you said, that would allow you to share the instance with other libraries.

I'm refactoring the application to include tests (to improve stability) and will fix this issue after that. 

Cheers for reporting this, 
Ben

**Comment by tfesslerCWC** [comment]
Thanks..  Much appreciated.   This will help down the road. 

If you ever consider integrating a file upload platform, please don't follow the existing examples for react-admin and firebase that leverages image conversion to a base 64 string stored in firestore. (shaking my head no)... It's a super-inefficient way to bloat firestore usage and costs.

With a shared Firebase instance, I highly recommend blending the react-firebase-filepond repo.  It offers a good solution for a single, managed, upload instance that can be easily modified to manage multiple content folders with ease.  

On a side note... when you are adding the lastmodified date to an item, it would be great to know the users email address that's making the change.    While it's helpful to know when a change was made, it helps to KNOW who made the change that potentially broke the system, helping to avoid.. the it wasn't me syndrome that users seem to take on.

**Comment by benwinding** [comment]
This issue is fixed in version `0.6.2`. Example:
``` javascript
const options = {
  app: firebaseAppInstance
}
const authProvider = FirebaseAuthProvider(firebaseConfig, options);
const dataProvider = FirebaseDataProvider(firebaseConfig, options);
const firebaseRealtime = FirebaseRealTimeSaga(dataProvider, options);
```
Have a go and let me know!

The upload feature is on the horizon, and yes the base64 solution is easy but ugly, so I don't think I'll use that, firebase storage buckets would be the way to go! 

Also, the "modified by who" field would be really great and could be implemented fairly easily, on the horizon too.

Cheers,
Ben

**Comment by tfesslerCWC** [comment]
Great timing!!   I've been working with the a refactor of react-firebase-filepond to handle image uploading.    While the default example uploads to one folder.  This is were react-admin-firebase can make it much more powerful...

Here's a quick overview of the approach..
I've created a admin folders set of components: list, create, edit, and show. 

**on the show screen**
I placed the  react-filepond instance - it provides a drag and dro/system dialog box upload interface, showing the contents of the folder name from a separate database child reference.  In this manner, one can have multiple upload folder destinations for the files.

**with react-filepond:**
- images are uploaded to firebase storage with a path of  /images/foldername/{filename}.
- files displayed in folder show are tracked in a database path of images/foldername/files/{filename with illegal chars removed)/metadata
- logs are tracked in images/foldername/logs.
- for each file that's uploaded, one last record is added to a cdn-sync database table. 

**with firebase functions**
- every time a folder is created in the db, upload folders are added to the cdn.
- a firebase function  runs every 5 minutes to coordinate the transfer of the storage files in the queue to CDN backed storage.

**Comment by benwinding** [comment]
Thanks for putting some work into this @tfesslerCWC,

It's clear that using google-storage-buckets is not a straight forward exercise, as you've certainly discovered!

In our company we use firebase and have also been through several approaches while developing our upload component. We decided to steer away from having _state_ tracked in the database, and extra cloud functions to handle uploads. The approach we have taken is as follows:

### Plain old File `<input />` field
- Styled as a `dropzone` type component
- Has drag and drop capabilities in all browsers
- When dropping a file to this component On drop initiates an upload to storage based using the firebase client sdk `app.storage()`
- A reference to that storage item _and_ the download_url is returned and stored in the firestore document's field.

### Storage management
- Files uploaded are stored in the default firebase bucket.
- They are located in the bucket under a path which matches the firestore document you're attaching the File to e.g.; `/users/089uasd9asd/profile_pic/image1.png`
- Deleting a file can be done easily through the firebase client sdk and the reference removed from the firestore document.
- If a document is deleted, you can simply delete all associated files under that storage location e.g: `remove all files under storage path: /users/089uasd9asd`

We decided to use this approach because Firestore document paths are _immutable_ so the reference to the storage path will always match the database (as both cannot be changed, only deleted). We also use the firebase storage bucket and immediately generate the public_url (or CDN url), which using google's CDN, not sure if that's what you're doing.

This approach is very easy to implement for the end user as it doesn't require any more database tracking collections or periodic cloud functions, but could be difficult in some unforeseen ways.

Let me know what you think!

Cheers,
Ben

**Comment by tfesslerCWC** [comment]
Thanks for sharing your approach...  I'm looking forward to dropping my in-progress solution into the updated react-admin-firebase release on Monday.   

I'm building a system to manage content for multiple users from multiple business units working across multiple shared project folders with anywhere from 50-250+ files.   Content maintained on Firebase Storage is auto-deployed to pre-production environments via secure file transfer.  As needed, content can be promoted/restored to any environment on an  immediate or timed deployment on a per file / per project basis.   While I would like to use one upload container path, it's not an option and tracking the state of each file across multiple environments is critically necessary.   Also, the need to promote to production on a timed basis forced me to go down the functions rabbit hole and where it ended up making sense to develop all of the promotion coordination between storage and the cdn. 

While deleting content was easy to implement, I'm concerned with maintaining the integrity of production environments because humans make human mistakes.    After several days of thought I'm going to restrict users to 'marking content for deletion' where they have time to go back and fix a mistake before it happens.  

Out of the box, Firebase Storage is not backed by cached CDN and uploaded content does not have any cache-control headers.  Are you setting any cache-control parameters of the images on upload to Firebase Storage?   Unfortunately Firebase storage caching !==  Firebase hosting caching.  While there are complex methods to leverage functions to provide hosting style caching, it's something I haven't explored yet.

**Comment by benwinding** [comment]
Cheers for detailing your process a bit more, I think I'm getting a clearer picture of what you mean now. It sounds like your kind of syncing the files from firebase storage to firebase hosting (which is your CDN) right?

That's got the advantage of CDN caching which is nice, but from what I've seen firebase deploy's to hosting are _atomic deploys_, so don't you need to reupload _all_ files every time you update firebase hosting? or are you using a different CDN?

Firebase storage seems like a good option to provide a consistent state of what's been upload, but caching is definitely not as easy as firebase hosting.

Cheers again,
Ben

**Comment by tfesslerCWC** [comment]
I've take a look at the updated structure and it's close.... But not quite there yet...

Approach implemented in 0.6.1:
(config.json - contains the firebase config ) - > App.JS -> FirebaseDataProvider(config.json, options)

Needed:
(config.js - contains the config and exports an initialized Firebase App)  ->  App.js ->  FirebaseDataProvider(app, options)

The initial request, was to allow for the initialization of the Firebase instance on the App layer, that can be shared with the React-Admin-Firebase library.   In the 'Needed' example, the initialized app is passed to the ReactAdmin Library, and can also be easily accessed for other uses in the client layer.

**Comment by benwinding** [comment]
Sorry @tfesslerCWC,

The docs, haven't quiet explained this well enough, if you pass `null` to the Dataprovider and put your app instance in the options, then the `react-admin-firebase` system will use _your_ instance instead of initializing one. (example below)

``` javascript
import {firebaseAppInstance} from  'config.js'

const options = {
  // Your own, previously initialized firebase app instance
  app: firebaseAppInstance;
}

const dataProvider = FirebaseDataProvider(null, options);
const authProvider = FirebaseAuthProvider(null, options);
const firebaseRealtime = FirebaseRealTimeSaga(dataProvider, options);
```

The null reference was necessary for backwards compatibility. It may be removed in future versions, so everything is in the options object.

Let me know if this is still what you were thinking.

Ben
